fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/derby/ReindexResourceDAO.java

```diff
@@ -0,0 +1,292 @@
+import com.ibm.fhir.persistence.exception.FHIRPersistenceResourceDeletedException;
+import com.ibm.fhir.persistence.exception.FHIRPersistenceResourceNotFoundException;
+import com.ibm.fhir.persistence.jdbc.FHIRPersistenceJDBCCache;
+import com.ibm.fhir.persistence.jdbc.connection.FHIRDbFlavor;
+import com.ibm.fhir.persistence.jdbc.dao.api.IResourceReferenceDAO;
+import com.ibm.fhir.persistence.jdbc.dao.api.JDBCIdentityCache;
+import com.ibm.fhir.persistence.jdbc.dao.api.ParameterDAO;
+import com.ibm.fhir.persistence.jdbc.dao.api.ResourceIndexRecord;
+import com.ibm.fhir.persistence.jdbc.dao.impl.JDBCIdentityCacheImpl;
+import com.ibm.fhir.persistence.jdbc.dao.impl.ParameterVisitorBatchDAO;
+import com.ibm.fhir.persistence.jdbc.dao.impl.ResourceDAOImpl;
+import com.ibm.fhir.persistence.jdbc.dto.ExtractedParameterValue;
+import com.ibm.fhir.persistence.jdbc.impl.ParameterTransactionDataImpl;
+
+/**
+ * DAO used to contain the logic required to reindex a given resource
+ */
+public class ReindexResourceDAO extends ResourceDAOImpl {
+    private static final Logger logger = Logger.getLogger(ReindexResourceDAO.class.getName());
+    private static final String CLASSNAME = ReindexResourceDAO.class.getSimpleName();
+    private static final SecureRandom random = new SecureRandom();
+
+    // The translator specific to the database type we're working with
+    private final IDatabaseTranslator translator;
+    
+    private final ParameterDAO parameterDao;
+
+    /**
+     * Public constructor
+     * @param connection
+     * @param translator
+     * @param parameterDao
+     * @param schemaName
+     * @param flavor
+     * @param cache
+     * @param rrd
+     */
+    public ReindexResourceDAO(Connection connection, IDatabaseTranslator translator, ParameterDAO parameterDao, String schemaName, FHIRDbFlavor flavor, FHIRPersistenceJDBCCache cache, IResourceReferenceDAO rrd) {
+        super(connection, schemaName, flavor, cache, rrd);
+        this.translator = translator;
+        this.parameterDao = parameterDao;
+    }
+
+    /**
+     * Public constructor
+     * @param connection
+     * @param translator
+     * @param parameterDao
+     * @param schemaName
+     * @param flavor
+     * @param trxSynchRegistry
+     * @param cache
+     * @param rrd
+     */
+    public ReindexResourceDAO(Connection connection, IDatabaseTranslator translator, ParameterDAO parameterDao, String schemaName, FHIRDbFlavor flavor, TransactionSynchronizationRegistry trxSynchRegistry, FHIRPersistenceJDBCCache cache, IResourceReferenceDAO rrd, ParameterTransactionDataImpl ptdi) {
+        super(connection, schemaName, flavor, trxSynchRegistry, cache, rrd, ptdi);
+        this.translator = translator;
+        this.parameterDao = parameterDao;
+
+    }
+
+    /**
+     * Getter for the translator currently held by this DAO
+     * @return
+     */
+    protected IDatabaseTranslator getTranslator() {
+        return this.translator;
+    }
```

Comment:
When is this important?

Reply:
For subclasses like the Postgres and Derby specializations.